cupy/_core/_kernel.pyx

```diff
@@ -874,14 +874,33 @@ cdef class ElementwiseKernel:
         return kern
 
 
+cdef str fix_cast_expr(src_type, dst_type, str expr):
+    src_kind = get_dtype(src_type).kind
+    dst_kind = get_dtype(dst_type).kind
+    if src_kind == dst_kind:
+        return expr
+    if src_kind == 'b':
+        return f'({expr}) ? 1 : 0'
+    if src_kind == 'c':
+        return f'({expr}).real()'
+    return expr
+
 cdef function.Function _get_ufunc_kernel(
-        tuple in_types, tuple out_types, routine, tuple arginfos, params,
+        tuple in_types, tuple out_types, routine, tuple arginfos,
+        bint has_where, params,
```

Comment:
note: (internal discussion with @toslunar)
`has_where` argument can be removed by introducing a **constant array**, that is a ndaray-like compile-time constant object.
If `where` option is not given, `where` becomes a constant `True`.